### PR TITLE
Implement skip zeroed blocks

### PR DESCRIPTION
**Pull request by Nefelim4ag**
Signed-off-by: Timofey Titovets <nefelim4ag@gmail.com>

### CODE REVIEW

**Comment by Nefelim4ag** [comment]
```
truncate -s 8G 8G_{1..4}
➜  test time sudo duperemove -vhrd ~/test
Using 128K blocks
Using hash: murmur3
Using 4 threads for file hashing phase
csum: /home/nefelim4ag/test/8G_1        [1/4] (25.00%)
csum: /home/nefelim4ag/test/8G_2        [2/4] (50.00%)
csum: /home/nefelim4ag/test/8G_3        [3/4] (75.00%)
csum: /home/nefelim4ag/test/8G_4        [4/4] (100.00%)
Total files:  4
Total hashes: 262144
Loading only duplicated hashes from hashfile.
Hashing completed. Calculating duplicate extents - this may take some time.
Gather files from hash: e47862ea (262144 identical extents)
Process 4 files.
[4] Compare files "/home/nefelim4ag/test/8G_2" and "/home/nefelim4ag/test/8G_1"
[4] Compare files "/home/nefelim4ag/test/8G_2" and "/home/nefelim4ag/test/8G_4"
[4] Compare files "/home/nefelim4ag/test/8G_2" and "/home/nefelim4ag/test/8G_3"
[3] Compare files "/home/nefelim4ag/test/8G_1" and "/home/nefelim4ag/test/8G_4"
[3] Compare files "/home/nefelim4ag/test/8G_1" and "/home/nefelim4ag/test/8G_3"
[2] Compare files "/home/nefelim4ag/test/8G_4" and "/home/nefelim4ag/test/8G_3"
Removing overlapping extents
Simple read and compare of file data found 1 instances of extents that might benefit from deduplication.
Showing 4 identical extents with id 2b7f489d
Start           Length          Filename
0.0     8.0G    "/home/nefelim4ag/test/8G_2"
0.0     8.0G    "/home/nefelim4ag/test/8G_1"
0.0     8.0G    "/home/nefelim4ag/test/8G_4"
0.0     8.0G    "/home/nefelim4ag/test/8G_3"
Using 4 threads for dedupe phase
[0x23b7590] Try to dedupe extents with id 2b7f489d
[0x23b7590] Add extent for file "/home/nefelim4ag/test/8G_2" at offset 0.0
[0x23b7590] Add extent for file "/home/nefelim4ag/test/8G_1" at offset 0.0
[0x23b7590] Add extent for file "/home/nefelim4ag/test/8G_4" at offset 0.0
[0x23b7590] Add extent for file "/home/nefelim4ag/test/8G_3" at offset 0.0
[0x23b7590] Dedupe 3 extents (id: 2b7f489d) with target: (0.0, 8.0G), "/home/nefelim4ag/test/8G_2"
Kernel processed data (excludes target files): 24.0G
Comparison of extent info shows a net change in shared extents of: 0.0
sudo duperemove -vhrd ~/test  8.43s user 69.82s system 155% cpu 50.450 total

➜  test time sudo ~/Coding/duperemove/duperemove --skip-zeroes -vhrd ~/test
Using 128K blocks
Using hash: murmur3
Using 4 threads for file hashing phase
csum: /home/nefelim4ag/test/8G_1        [1/4] (25.00%)
csum: /home/nefelim4ag/test/8G_2        [2/4] (50.00%)
csum: /home/nefelim4ag/test/8G_3        [3/4] (75.00%)
csum: /home/nefelim4ag/test/8G_4        [4/4] (100.00%)
Total files:  4
Total hashes: 0
Loading only duplicated hashes from hashfile.
Hashing completed. Calculating duplicate extents - this may take some time.

Search completed with no errors.             
Simple read and compare of file data found 0 instances of extents that might benefit from deduplication.
sudo ~/Coding/duperemove/duperemove --skip-zeroes -vhrd ~/test  3.47s user 33.96s system 222% cpu 16.825 total
```

**Comment by clobrother** [comment]
Test it with mixed files, will it work ?

**Comment by Nefelim4ag** [comment]
@clobrother, yes, i've also test it on RAW VM images

File: duperemove.c [hunk 1]
```diff
@@ -183,6 +186,7 @@ static int parse_options(int argc, char **argv)
 		{ "lookup-extents", 1, NULL, LOOKUP_EXTENTS_OPTION },
 		{ "one-file-system", 0, NULL, ONE_FILESYSTEM_OPTION },
 		{ "hash", 1, NULL, HASH_OPTION },
+		{ "skip-zeroes", 1, NULL, SKIP_ZEROES_OPTION },
```
Comment (markfasheh) [inline duperemove.c]: Want to do a '0' for the 2nd field here as skip-zeroes does not take an argument. This is probably why it only worked one way and not the other :)

**Comment by markfasheh** [comment]
Thanks for this - instead of is_block_zeroed() can you just take the fiemap result and decide based on whether that code returns a hole or not? You might need to move that small bit of code up in csum_next_block() but otherwise that would avoid the extra read completely.

Or am I mistaken and is this only about skipping allocated but zero'd blocks?  If so remind me why we want this please  :)

**Comment by Nefelim4ag** [comment]
@markfasheh, thanks - copy past error was fixed %)
---
I tried it, but this if is always false (fc == NULL):
```
	if (fc) {
		unsigned int fieflags = 0;

		ret = fiemap_iter_get_flags(fc, data->file, *off, &fieflags,
					    &hole);
```
So, if i can't fix it, i decide to implement it more simply %)
---
I implement it for speedup analyzing phase, not matter why zeroes are happen:
By punch a hole, or by file zeroing, it's will take extra time when we work with it
Example: one of VM store 7.0 TB (du --apparent-size 12 TB) of data (some of them have zeroed blocks)
I don't want from duperemove to spend time and analyze zeroes, i can cover zeroes with fs compression

**Comment by markfasheh** [comment]
Ahh it was false because I'm hiding it behind the --lookup-extents option. If you check csum_whole_file_init() we do the allocation of fiemap_context there.

It's off by default because we used to try doing fancy things with the SHARED flag which never worked out. Right now however, it's set then ignored. If you want, you can change do_lookup_extents be on by default (in a separate commit please). That'll give you the context and from there you can just skip the rest of the function when we see a hole. Otherwise I'd suggest doing --lookup-extents=yes in your test :)

It might actually be good to also keep your --skip-zeroes code around too though. Skipping holes and detecting/skipping allocated zero'd blocks have slightly different applications. In that case though it seems like I might be asking you to do some extra work. If that's too much don't worry about, I can pull this once I have the chance to sit down and make the 'skip holes' changes myself. :)

Thanks!

**Comment by Nefelim4ag** [comment]
@markfasheh, i've try to implement it, but hole detection not work properly =\
(especially for files with no extent (i.e. files created by truncate))

This code is never true in my tests
```
	if (blkno < extent->fe_logical) {
		*hole = 1;
		return 0;
	}
```

And fieflags haven't flag for holes
- - -
Some test files:
```
# truncate -s 4M full_hole
# truncate -s 1G with_holes; losetup -f with_holes; mkfs.ext4 /dev/loop0; mount /dev/loop0 /mnt; dd if=/dev/zero of=/mnt/zero; rm /mnt/zero; dd if=/dev/zero of=/mnt/zero bs=1M count=64; fstrim /mnt;
```

As i see in my tests:
For truncated files (but detect only first block):
```
--- a/filerec.c
+++ b/filerec.c
@@ -487,6 +487,8 @@ int fiemap_iter_get_flags(struct fiemap_ctxt *ctxt, struct filerec *file,
 
        if (ctxt->idx == -1) {
                err = do_fiemap(fiemap, file, 0);
+               if (fiemap->fm_mapped_extents == 0)
+                       *hole = 1;
                if (err || fiemap->fm_mapped_extents == 0)
                        goto out_last_map;
```

for files with holes
```
diff --git a/filerec.c b/filerec.c
index 45c377c..e727836 100644
--- a/filerec.c
+++ b/filerec.c
@@ -495,6 +495,8 @@ int fiemap_iter_get_flags(struct fiemap_ctxt *ctxt, struct filerec *file,
 
 check:
        extent = &fiemap->fm_extents[ctxt->idx];
+       if (blkno == extent->fe_logical)
+               *hole = 1;
        if (block_contained(blkno, extent)) {
                *flags = extent->fe_flags;
                return 0;
```

- - -
So i can't fix hole problem detection, sorry.

**Comment by markfasheh** [comment]
Oh ok so we're not catching the case where there's a hole between isize and end of the allocation. Thanks, I'll give it some review and hopefully a pull.

**Pull request merged**